### PR TITLE
[CPP Speech Sample] Improve `-o` and `-oname` flags

### PR DESCRIPTION
**Pull request by dpigasin**
### Details:
 - TODO: refactoring

### CODE REVIEW

**Comment by vladimir-dudnik** [comment]
@dpigasin some precommit check failed

**Comment by vladimir-dudnik** [comment]
some tests failed:

00:26:27  [2022-02-14 21:27:12,402] [140217595279104] samples_ie_tests-0 INFO: [2022-02-14 21:27:12,402] [139795149459712] helper.cmd_exec INFO: _ TestSpeechSample.test_speech_sample_oname_cpu[{'m': 'wsj/wsj_dnn5b_smbr_v10_fp32_fused_batch_1.xml', 'i': 'ark/dev93_10.ark', 'r': 'ark/dev93_10.ark', 'd': 'CPU', 'oname': 'Parameter:0', 'bs': 1, 'sample_type': 'C++'}] _
00:26:27  [2022-02-14 21:27:12,402] [140217595279104] samples_ie_tests-0 INFO: [2022-02-14 21:27:12,402] [139795149459712] helper.cmd_exec INFO: Traceback (most recent call last):
00:26:27  [2022-02-14 21:27:12,402] [140217595279104] samples_ie_tests-0 INFO: [2022-02-14 21:27:12,402] [139795149459712] helper.cmd_exec INFO:   File "/home/jenkins/agent/workspace/private-ci/ie/e2e-tests-linux-ubuntu20-samples/b/repos/frameworks.ai.openvino.tests/sample_smoke_test/ie_samples/test_speech_sample.py", line 428, in test_speech_sample_oname_cpu
00:26:27  [2022-02-14 21:27:12,402] [140217595279104] samples_ie_tests-0 INFO: [2022-02-14 21:27:12,402] [139795149459712] helper.cmd_exec INFO:     stdout = self._test(param).split('\n')
00:26:27  [2022-02-14 21:27:12,402] [140217595279104] samples_ie_tests-0 INFO: [2022-02-14 21:27:12,402] [139795149459712] helper.cmd_exec INFO:   File "/home/jenkins/agent/workspace/private-ci/ie/e2e-tests-linux-ubuntu20-samples/b/repos/frameworks.ai.openvino.tests/sample_smoke_test/common/samples_common_test_class.py", line 440, in _test
00:26:27  [2022-02-14 21:27:12,402] [140217595279104] samples_ie_tests-0 INFO: [2022-02-14 21:27:12,402] [139795149459712] helper.cmd_exec INFO:     assert retcode == 0, "Sample execution failed. Exit code is {}".format(retcode)
00:26:27  [2022-02-14 21:27:12,403] [140217595279104] samples_ie_tests-0 INFO: [2022-02-14 21:27:12,402] [139795149459712] helper.cmd_exec INFO: AssertionError: Sample execution failed. Exit code is 1

File: samples/cpp/speech_sample/main.cpp [hunk 1]
```diff
@@ -286,161 +298,182 @@ int main(int argc, char* argv[]) {
...
-            // initialize memory state before starting
-            for (auto&& state : inferRequests.begin()->inferRequest.query_state()) {
-                state.reset();
+        // --------------------------- Step 5. Do inference
+        // --------------------------------------------------------
+        std::vector<std::vector<uint8_t>> ptrUtterances;
+        std::vector<std::vector<uint8_t>> vectorPtrScores((outputs.size() == 0) ? 1 : outputs.size());
+        std::vector<uint8_t> ptrScores;
+        std::vector<uint16_t> numScoresPerOutput((outputs.size() == 0) ? 1 : outputs.size());
+        std::vector<uint8_t> ptrReferenceScores;
```
[the remark this answers is not in the document]
Comment (nkogteva) [reply samples/cpp/speech_sample/main.cpp:321]: BTW it is easy introduce map with name as key in scope of 77987 ticket  instead of having vector of vectors

File: samples/cpp/speech_sample/main.cpp [hunk 2]
```diff
@@ -286,161 +298,182 @@ int main(int argc, char* argv[]) {
...
+                uint32_t currentNumFrames(0), currentNumFrameElementsInput(0), currentNumBytesPerElementInput(0);
+                file->get_file_info(inputFilename, utteranceIndex, &n, &numBytesThisUtterance[i]);
+                ptrUtterance.resize(numBytesThisUtterance[i]);
+                file->load_file(inputFilename,
+                                utteranceIndex,
+                                uttName,
+                                ptrUtterance,
+                                &currentNumFrames,
+                                &currentNumFrameElementsInput,
+                                &currentNumBytesPerElementInput);
```
Comment (nkogteva) [inline samples/cpp/speech_sample/main.cpp:352]: It would be nice to introduce structure to work with files. The same things are doing several times for input/output/reference files. Too many variables which is hard to track.

For example:
```
        struct UtteranceInfo {
            uint32_t num_frames = 0;            // num_rows
            uint32_t num_frame_elements = 0;    // num_cols
            uint32_t num_bytes_per_element = 0; // elem_size
            uint32_t num_bytes = 0;             // num_rows * num_cols * elem_size
            std::vector<uint8_t> data = {};
            uint8_t* data_ptr = nullptr;
        };

        struct FileInfo {
            std::string path = "";
            BaseFile* file_ptr = nullptr;
            uint32_t num_utterances = 0;
            uint32_t num_memory_bytes = 0;
            std::vector<UtteranceInfo> u_info = {};
        };
```

File: samples/cpp/speech_sample/main.cpp [hunk 3]
```diff
@@ -286,161 +298,182 @@ int main(int argc, char* argv[]) {
...
+            auto t0 = Time::now();
+            auto t1 = t0;
+
+            BaseFile* fileReferenceScores;
+            std::string refUtteranceName;
+
+            if (!FLAGS_r.empty()) {
+                /** Read file with reference scores **/
+                auto exReferenceScoresFile = fileExt(FLAGS_r);
+                if (exReferenceScoresFile == "ark") {
```
Comment (nkogteva) [inline samples/cpp/speech_sample/main.cpp:394]: Can be wrapped using function.

File: samples/cpp/speech_sample/main.cpp [hunk 4]
```diff
@@ -286,161 +298,182 @@ int main(int argc, char* argv[]) {
...
-                                                   refUtteranceName,
-                                                   ptrReferenceScores,
-                                                   &numFramesReference,
-                                                   &numFrameElementsReference,
-                                                   &numBytesPerElementReference);
+                                                       refUtteranceName,
+                                                       ptrReferenceScores,
+                                                       &numFramesReference,
+                                                       &numFrameElementsReference,
+                                                       &numBytesPerElementReference);
```
Comment (nkogteva) [inline samples/cpp/speech_sample/main.cpp:414]: We need a test for model with several outputs

**Review by vladimir-dudnik — APPROVED** [review]

File: samples/cpp/speech_sample/main.cpp [hunk 5]
```diff
@@ -32,16 +32,19 @@
...
+ * @brief The entry point for inference engine automatic speech recognition
+ * sample
  * @file speech_sample/main.cpp
  * @example speech_sample/main.cpp
  */
 int main(int argc, char* argv[]) {
     try {
-        // ------------------------------ Get Inference Engine version ----------------------------------------------
+        // ------------------------------ Get Inference Engine version
+        // ----------------------------------------------
```
[the remark this answers is not in the document]
Comment (dpigasin) [reply samples/cpp/speech_sample/main.cpp]: fixed

File: samples/cpp/speech_sample/main.cpp [hunk 6]
```diff
@@ -286,161 +298,182 @@ int main(int argc, char* argv[]) {
...
-            ScoreErrorT frameError, totalError;
-            ptrUtterances.resize(inputFiles.size());
-            // initialize memory state before starting
-            for (auto&& state : inferRequests.begin()->inferRequest.query_state()) {
-                state.reset();
+        // --------------------------- Step 5. Do inference
+        // --------------------------------------------------------
+        std::vector<std::vector<uint8_t>> ptrUtterances;
+        std::vector<std::vector<uint8_t>> vectorPtrScores((outputs.size() == 0) ? 1 : outputs.size());
+        std::vector<uint8_t> ptrScores;
```
[the remark this answers is not in the document]
Comment (dpigasin) [reply samples/cpp/speech_sample/main.cpp]: `std::vector<uint8_t> ptrScores;` was removed

File: samples/cpp/speech_sample/main.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (dpigasin) [reply samples/cpp/speech_sample/main.cpp]: `ptrReferenceScores` is vector now. Maybe we change it to map in future.

File: samples/cpp/speech_sample/main.cpp (hunk 4 above)
Comment (dpigasin) [reply samples/cpp/speech_sample/main.cpp]: Now we make a comparison with reference for each output.

**Comment by vladimir-dudnik** [comment]
there are failed tests:

/home/jenkins/agent/workspace/private-ci/ie/e2e-tests-linux-ubuntu20-samples/b/install/dldt_cpack/samples_bin/speech_sample -bs 8 -d CPU -i /nfs/inn/proj/vdp/vdp_tests/samples_tests_data/ark/aspire_tdnn/mini_feat_1_10.ark,/nfs/inn/proj/vdp/vdp_tests/samples_tests_data/ark/aspire_tdnn/mini_feat_1_10_ivector.ark -iname input,ivector -layout [NC] -m /nfs/inn/proj/vdp/vdp_tests/samples_tests_data/private_ir/aspire_tdnn/aspire_tdnn_fp32.xml -r /nfs/inn/proj/vdp/vdp_tests/samples_tests_data/ark/aspire_tdnn/aspire_tdnn_mini_feat_1_10_kaldi_score.ark 
23:04:34  [2022-02-18 20:04:33,664] [140098393237248] samples_ie_tests-0 INFO: [2022-02-18 20:04:33,664] [140075031163136] helper.cmd_exec INFO: [ ERROR ] [ ERROR ] Number of output and reference frames does not match.

**Review by nkogteva — APPROVED** [review]

File: samples/cpp/speech_sample/main.cpp [hunk 7]
```diff
@@ -110,8 +111,7 @@ int main(int argc, char* argv[]) {
             model = core.read_model(FLAGS_m);
             if (!outputs.empty()) {
                 for (size_t i = 0; i < outputs.size(); i++) {
-                    auto output = model->add_output(outputs[i], ports[i]);
-                    output.set_names({outputs[i] + ":" + std::to_string(ports[i])});
```
Comment (mryzhov) [inline samples/cpp/speech_sample/main.cpp:114]: Why did you remove this string? Did you test the speech sample with the new output?

Comment (dpigasin) [reply samples/cpp/speech_sample/main.cpp:114]: Outputs of IR10 and IR11 models already have name with `:`

Comment (mryzhov) [reply samples/cpp/speech_sample/main.cpp:114]: But other layers - no.
Did you perform the following tests:
1) set any layers' output as results (--oname="layer_name:layer_port")
2) for the networks with several outputs get the result from the one only

Comment (mryzhov) [reply samples/cpp/speech_sample/main.cpp:114]: > I tested it with `-oname sigmoid5:0,Parameter:0`. The names of outputs was `sigmoid5:0` and `Parameter:0`

Because Parameter has tensor name too as input node :)

**Review by mryzhov — APPROVED** [review]
Creating of several score files causes sample crash ("--o", "cpu_1.ark,cpu_2.ark,cpu_3.ark")

**Comment by dpigasin** [comment]
> Creating of several score files causes sample crash ("--o", "cpu_1.ark,cpu_2.ark,cpu_3.ark")

Can't reproduce with the public model (wsj_dnn5b_smbr)
